cdap-master/src/main/java/co/cask/cdap/data/runtime/main/MetricsProcessorTwillRunnable.java

```diff
@@ -136,20 +137,30 @@ protected void configure() {
   }
 
   static final class MetricsProcessorModule extends PrivateModule {
+    final Integer instanceId;
+
+    MetricsProcessorModule(TwillContext twillContext) {
+      this.instanceId = twillContext.getInstanceId();
+    }
+
     @Override
     protected void configure() {
       // Single MetricDatasetFactory used by both KafkaMetricsProcessorService and MessagingMetricsProcessorService
       bind(MetricDatasetFactory.class).to(DefaultMetricDatasetFactory.class).in(Scopes.SINGLETON);
       bind(MessageCallbackFactory.class).to(MetricsMessageCallbackFactory.class);
+      bind(Integer.class).annotatedWith(Names.named(Constants.Metrics.TWILL_INSTACE_ID)).toInstance(instanceId);
+
       install(new FactoryModuleBuilder()
                 .build(KafkaMetricsProcessorServiceFactory.class));
 
-      expose(KafkaMetricsProcessorServiceFactory.class);
-
       install(new FactoryModuleBuilder()
                 .build(MessagingMetricsProcessorServiceFactory.class));
 
-      expose(MessagingMetricsProcessorServiceFactory.class);
+      bind(MessagingMetricsProcessorRuntimeService.class);
```

Comment:
otherwise it would complain, `Could not expose() co.cask.cdap.metrics.runtime.MessagingMetricsProcessorRuntimeService, it must be explicitly bound.`